File: public/tools/elements/app/GtView.ts
import {PolymerElement} from "../../polymer/PolymerElement";
import {Element, Inject, On, PolymerElementDeclaration} from "../../polymer/Annotations";
import {ViewMetadata} from "../../client/router/View";
import {Constructor} from "../../utils/DI";
import {Router} from "../../client/router/Router";
import {Loader} from "../../client/loader/Loader";

// Dummy factory div to construct views instances
const factory = document.createElement("div");

@Element({
	selector: "gt-view",
	template: "/assets/imports/app.html"
})
export class GtView extends PolymerElement {
	@Inject
	private loader: Loader;

	@Inject
	@On({"route-updated": "update"})
	private router: Router;
	private last_view: Constructor<PolymerElement> = null;

	private layer: number = 0;
	private current: any = null;

	private async update() {
		let new_view = this.router.activeView;
		let args = this.router.activeArguments;

		if (new_view != this.last_view) {
			document.body.classList.add("app-loader");
		}

		let meta = new_view ? Reflect.getMetadata<ViewMetadata>("view:meta", new_view) : null;

		// If the view is the same and is sticky, do not remove the
		// current element but update attributes values
		if (new_view == this.last_view && meta && meta.sticky) {
			for (let key in args) {
				this.current.setAttribute(key, args[key] != void 0 ? args[key] : null);
			}
			return;
		}

		this.current = null;

		// Remove every current children
		let views: Element[] = this.node.children;
		let tasks: Promise<any>[] = views.map(e => {
			return new Promise((res) => {
				e.classList.remove("active");
				e.addEventListener("transitionend", res);
			}).then(() => this.node.removeChild(e));
		});

		this.last_view = new_view;
		if (!new_view) return;

		// Prevent navigation during the transition
		this.router.lock(true);

		// Wait until everything is ready to create the view element
		tasks.push(this.loader.loadElement(new_view));
		await Promise.all(tasks);

		// Hide loader
		document.body.classList.remove("app-loader");

		// Construct the argument list
		let arg_string: string[] = [];
		for (let key in args) {
			if (args[key] != void 0) {
				let arg_value = args[key].replace(/./g, (s) => `&#${s.charCodeAt(0)};`);
				arg_string.push(` ${key}='${arg_value}'`);
			}
		}

		// Use a crazy HTML generation system to create the element since we need to have
		// router-provided attributes defined before the createdCallback() method is called
		let decl = Reflect.getMetadata<PolymerElementDeclaration>("polymer:declaration", new_view);
		if (!decl) return;
		factory.innerHTML = `<${decl.selector}${arg_string.join()}></${decl.selector}>`;

		// Element has been constructed by the HTML parser
		let element = <any> factory.firstElementChild;

		// Ensure the new element is over older ones no matter what
		element.style.zIndex = ++this.layer;

		// Insert the element in the DOM
		this.current = element;
		this.node.appendChild(element);

		// Add active class two frames from now
		requestAnimationFrame(() => requestAnimationFrame(() => element.classList.add("active")));

		// Unlock router and allow navigation again
		this.router.lock(false);
	}
}

File: client/src/gt/component/app/GtView.scala
package gt.component.app

import gt.component.{GtHandler, View}
import gt.{App, Router}
import org.scalajs.dom._
import org.scalajs.dom.raw.HTMLElement
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.{Future, Promise}
import scala.util.{Failure, Success}
import util.implicits._
import util.jsannotation.js
import xuen.Component

object GtView extends Component[GtView](
	selector = "gt-view",
	templateUrl = "/assets/imports/app.html"
)

@js class GtView extends GtHandler {
	private[this] var currentLayer = 0
	private[this] var currentNode: Element = null
	private[this] var currentArgs: Seq[(String, String)] = Nil
	private[this] var currentSelector: String = null

	Router.onUpdate ~> { case (_, args, view) => update(args, view) }

	private def update(args: Seq[(String, String)], view: View): Unit = {
		if (view.selector != currentSelector) {
			document.body.classList.add("app-loader")
		}

		// If the view is the same and is sticky, do not remove the current
		// element but update attributes values
		if (view.selector == currentSelector && view.sticky) {
			var map = args.toMap
			for ((name, _) <- currentArgs.toMap) map.get(name) match {
				case Some(value) =>
					currentNode.setAttribute(name, value)
					map = map - name
				case None =>
					currentNode.removeAttribute(name)
			}
			for ((name, value) <- args) {
				currentNode.setAttribute(name, value)
			}

			currentArgs = args
			return
		}

		currentNode = null

		// Remove every element children
		var tasks: Seq[Future[_]] = for (child <- children) yield {
			val promise = Promise[Unit]()
			child.classList.remove("active")
			child.addEventListener("transitionend", (e: Event) => {
				removeChild(child)
				promise.success(())
			})
			promise.future
		}

		// Add the element loading task
		tasks = tasks :+ view.component.load()

		// Prevent navigation during the transition
		Router.lock()

		// Wait until everything is ready to create the view element
		Future.sequence(tasks.asInstanceOf[Seq[Future[Unit]]]).onComplete {
			case Failure(e) =>
				window.console.error("Failure to load view", App.formatException(e))

			case Success(_) =>
				// Hide loader
				document.body.classList.remove("app-loader")

				// Construct the element
				val node = document.createElement(view.selector).asInstanceOf[HTMLElement]
				for ((name, value) <- args if value != null) node.setAttribute(name, value)

				// Ensure the new element is over older ones no matter what
				currentLayer += 1
				node.style.zIndex = currentLayer.toString

				currentNode = node
				currentArgs = args
				currentSelector = view.selector

				appendChild(node)

				// Add active class two frames from now
				// (don't ask why, I can't remember)
				window.requestAnimationFrame((_: Double) => {
					window.requestAnimationFrame((_: Double) => {
						node.classList.add("active")
					})
				})

				Router.unlock()
		}
	}
}
